Add get_min_timestamp to Bundle

nedgar opened the pull request:
Improves performance by taking timestamps from _StackedWindowedValues
rather than iterating over all elements.
Memoizes min timestamp if bundle is committed.

asfgit commented:
SUCCESS 
 
--none--

nedgar commented:
See thread at https://twitter.com/nedgar/status/929765848750977026
Basically, the min timestamp calculation in `_TransformWatermarks.refresh()` is expensive due to iterating over all elements in the bundle, and is called repeatedly. By adding `get_min_timestamp` to Bundle, it need only iterate over the root elements, e.g. taking timestamp directly from _StackedWindowedValues rather than iterating over their elements.
It also memoizes the min timestamp if the bundle is committed.

nedgar commented:
Before:
```
$ time python wordcount_minimal.py --input kinglear.txt --output counts
...
real	0m4.726s
user	0m4.457s
sys	0m0.214s

$ time python wordcount_minimal.py --input kinglear_x4.txt --output counts
...
real	0m22.268s
user	0m21.744s
sys	0m0.722s
```

After:
```
$ time python wordcount_minimal.py --input kinglear.txt --output counts
...
real	0m3.893s
user	0m3.623s
sys	0m0.189s

$ time python wordcount_minimal.py --input kinglear_x4.txt --output counts
...
real	0m10.994s
user	0m10.404s
sys	0m0.472s
```

Where `kinglear_x4.txt` is `kinglear.txt` repeated 4 times.

asfgit commented:
FAILURE
 
--none--

asfgit commented:
SUCCESS 
 
--none--

aaltay commented:
R: @charlesccychen

charlesccychen commented on sdks/python/apache_beam/runners/direct/bundle_factory.py:
```diff
@@ -198,3 +200,16 @@ def commit(self, synchronized_processing_time):
+
+  def get_min_timestamp(self):
+      """Returns the minimum timestamp across all elements in this bundle, or
+      MAX_TIMESTAMP if this bundle is empty."""
```
<!--new_thread; commit:61ae92dcffa98203cbe8f56930802405de2524f0; resolved:0-->
Can you make this one line?  For example:

```
   """Returns max element timestamp in bundle or MAX_TIMESTAMP if empty."""
```

charlesccychen commented on sdks/python/apache_beam/runners/direct/watermark_manager.py:
```diff
@@ -211,12 +211,9 @@ def refresh(self):
-          if wv.timestamp < min_pending_timestamp:
-            min_pending_timestamp = wv.timestamp
+        if input_bundle.has_elements():
+            min_pending_timestamp = min(min_pending_timestamp, input_bundle.get_min_timestamp())
```
<!--new_thread; commit:61ae92dcffa98203cbe8f56930802405de2524f0; resolved:0-->
Please use 2sp indentation, wrap line to 80 chars.

charlesccychen commented on sdks/python/apache_beam/runners/direct/bundle_factory.py:
```diff
@@ -198,3 +200,16 @@ def commit(self, synchronized_processing_time):
     self._elements = tuple(self._elements)
     self._synchronized_processing_time = synchronized_processing_time
+
+  def get_min_timestamp(self):
```
<!--new_thread; commit:61ae92dcffa98203cbe8f56930802405de2524f0; resolved:0-->
Please match indentation of the rest of the file (2sp).

nedgar commented:
Hi @charlesccychen. Here are some fixups addressing your review comments.
Should I merge with latest from apache/beam master, or will you?

Even with these changes, it takes several seconds to process even modest sized files (e.g. a few MB). Profiling suggests much of it is due to the creation of many Timestamp objects. Is the team aware of / planning to address this? Thanks.

charlesccychen commented:
Thank you!  Can you rebase to the head of master?

I am currently working on a rearchitecture of the DirectRunner to improve scalability, especially for multi-core environments (https://issues.apache.org/jira/browse/BEAM-1442).  This should be ready for use in the next few months.

charlesccychen commented:
LGTM

robertwb commented:
We are aware of the slow performance of the Direct runner, and hope to address it more directly. 

In the meantime, you could also try running with `pipeline(runner=apache_beam.runners.portability.fn_api_runner.FnApiRunner())` which can help on some pipelines.

nedgar commented:
I've rebased, thought not sure I've done it correctly. It looks OK in GitHub Desktop:

![image](https://user-images.githubusercontent.com/692834/33506536-4444bb88-d6be-11e7-9994-b3c17ddc64f3.png)

Thanks for the info re performance.

nedgar commented:
After the merge, running it fails due to missing `beam_runner_api_pb2`:
```
$ time python wordcount_minimal.py --input kinglear_x4.txt --output counts
Traceback (most recent call last):
  File "wordcount_minimal.py", line 53, in <module>
    import apache_beam as beam
  File "/Users/nedgar/Development/Apache-Beam/apache_beam/__init__.py", line 86, in <module>
    from apache_beam import coders
  File "/Users/nedgar/Development/Apache-Beam/apache_beam/coders/__init__.py", line 18, in <module>
    from apache_beam.coders.coders import *
  File "/Users/nedgar/Development/Apache-Beam/apache_beam/coders/coders.py", line 30, in <module>
    from apache_beam.portability.api import beam_runner_api_pb2
ImportError: cannot import name beam_runner_api_pb2
```

kennknowles commented:
Any updates here? It does look like the rebase was not quite right, as a good chunk of `apache:master` appears to have been rebased as well. You might explicitly rebase like so:

  `git rebase -i HEAD~2 origin/master`

Where I have chosen `HEAD~2` since it looks like you have two commits, and `-i` because you might as well squash them together at this point.

nedgar commented:
@kennknowles my apologies, I dropped the ball in getting back to this. I’m no longer exploring Beam, have switched jobs, including returning laptop with my changes. 
If you’d like to apply the proposed changes, please go ahead. 

I’m confused by the maven/gradle comments here. They seem out of context. Wrong PR or due to my bad merge?
